diff.go

```diff
@@ -0,0 +1,320 @@
+package amt
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"golang.org/x/xerrors"
+
+	"github.com/ipfs/go-cid"
+
+	cbor "github.com/ipfs/go-ipld-cbor"
+	cbg "github.com/whyrusleeping/cbor-gen"
+)
+
+// ChangeType denotes type of change in Change
+type ChangeType int
+
+// These constants define the changes that can be applied to a DAG.
+const (
+	Add ChangeType = iota
+	Remove
+	Modify
+)
+
+// Change represents a change to a DAG and contains a reference to the old and
+// new CIDs.
+type Change struct {
+	Type   ChangeType
+	Key    uint64
+	Before *cbg.Deferred
+	After  *cbg.Deferred
+}
+
+func (ch Change) String() string {
+	b, _ := json.Marshal(ch)
+	return string(b)
+}
+
+// Diff returns a set of changes that transform node 'a' into node 'b'.
+func Diff(ctx context.Context, prevBs, curBs cbor.IpldStore, prev, cur cid.Cid) ([]*Change, error) {
+	prevAmt, err := LoadAMT(ctx, prevBs, prev)
+	if err != nil {
+		return nil, xerrors.Errorf("loading previous root: %w", err)
+	}
+
+	curAmt, err := LoadAMT(ctx, curBs, cur)
+	if err != nil {
+		return nil, xerrors.Errorf("loading current root: %w", err)
+	}
+
+	return diffNode(ctx, prevBs, curBs, prevAmt.node, curAmt.node, int(prevAmt.height), int(curAmt.height), prevAmt.bitWidth, curAmt.bitWidth, 0)
+}
+
+func diffNode(ctx context.Context, prevBs, curBs cbor.IpldStore, prev, cur *node, prevHeight, curHeight int, prevBitWidth, curBitWidth uint, offset uint64) ([]*Change, error) {
+	if prev == nil && cur == nil {
+		return nil, nil
+	}
+
+	if prev == nil {
+		return addAll(ctx, curBs, cur, curBitWidth, curHeight, offset)
+	}
+
+	if cur == nil {
+		return removeAll(ctx, prevBs, prev, prevBitWidth, prevHeight, offset)
+	}
+
+	if prevHeight == 0 && curHeight == 0 {
+		return diffLeaves(prev, cur, offset)
+	}
+
+	changes := make([]*Change, 0)
+
+	if curHeight > prevHeight {
```

Comment:
Can we dedup this code with the code below?

Reply:
I looked and although they are structurally identical the numbers and roles of the parameters needed gives a shared function a pretty complex signature and doesn't save much code.

I simplified the function signatures as part of this but I don't see a good way to reduce it further at the moment.